doc/read-the-docs-site/adr/0005-plutus-tx-compiler-and-library.md

```diff
@@ -0,0 +1,209 @@
+What is unclear is whether Plutus Tx is strict or non-strict.
+Indeed, it is sometimes strict and sometimes non-strict, and it is fairly random and difficult to predict:
+
+- Let-bindings in Plutus Tx are by default non-strict, but function applications are strict.
+  That is, `let x = rhs in body` isn't equivalent to `(\x -> body) rhs`.
+  This is distinct from most other languages, regardless of their evaluation strategies, in which these two expressions are equivalent.
+
+- The conditional expression `if b then t else f` is always non-strict in `t` and `f`, like in most strict languages.
+  But what about `foo b t f = if b then t else f`? In a strict language, `foo` would be strict, and in a non-strict language, it would be non-strict.
+  But in Plutus Tx, due to how its compiler is currently implemented, whether `t` and `f` are evaluated strictly or not depends on whether GHC inlines `foo`.
+  Although we turn off GHC's inliner in the Plutus Tx compiler, GHC may still decide to unconditionally inline `foo` if it is used only once.
+  In general, it is not easy to predict whether GHC will inline something or not.
+
+The lack of consistency makes it difficult to reason about the efficiency of Plutus Tx programs, and difficult to write useful libraries for Plutus Tx that guarantee to have the desirable behaviors (e.g., short-circuiting).
+
+This ADR proposes a few changes to the Plutus Tx language and the compiler to make the semantics more predictable.
+It also proposes changes to the Plutus Tx standard library to make it more performant.
+
+## Proposed Changes
+
+Plutus Tx should be a strict-by-default language.
+Non-strict-by-default languages are rarely used in practice since non-strict evaluation leads to repeated evaluations.
+Unless we implement lazy evaluation for PLC in the future, strictness should be the default for Plutus Tx.
+Given this, the following changes are proposed.
+
+### Making Let-bindings Strict By Default
+
+In a strict-by-default language, it would make sense that let-bindings be strict by default, since that's what almost all strict languages do; plus, non-strict bindings should be used cautiously since they can easily lead to unbounded amount of repeated evaluation.
+But if make let-bindings strict by default, then the question is, how does one make a non-strict let-binding?
+
+Haskell supports lazy patterns with the `~` operator, and it can be used on let-bindings, e.g., `let ~x = rhs`. So one possibility is to make non-strict let-bindings this way.
+However, lazy pattern has no effect on simple pattern bindings (i.e., the pattern is a single variable), and it makes no difference in GHC Core, which means the Plutus Tx compiler can't actually see the `~`.
+It is technically possible to make `~` work but it would require adding a GHC parser plugin to capture `~` in GHC Core, among other things, adding a great deal of complexity.
```

Comment:
Can't we just enable the `Strict` pragma by default and then use `let ~x = rhs`? Quoting the [docs](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/strict.html#extension-Strict):

> Informally the `Strict` language extension switches functions, data types, and bindings to be strict by default, allowing optional laziness by adding `~` in front of a variable. This essentially reverses the present situation where laziness is default and strictness can be optionally had by adding `!` in front of a variable.

Reply:
Good question, I didn't consider `StrictData`. But how can we enable `StrictData` by default? It can probably be done in a DynFlags plugin, but that means using another plugin.

Reply:
Not `StrictData`, `Strict`. `Strict` is super weird and almost nobody uses it. I guess we could try it, but it would be invasive, and require changing the compilation flags of modules the plugin doesn't run on, perhaps :thinking: We could just say "we recommend using `Strict`" quite clearly and use it in our examples and libraries? Worth an experiment at least.